qutip/core/data/dispatch.pyx

```diff
@@ -552,8 +552,9 @@ cdef class Dispatcher:
         callable object which requires that the dispatched arguments match
         those specified in `types`.
         """
-        if isinstance(types, type):
+        if isinstance(types, (type, str)):
             types = (types,)
+        types = tuple(_to.parse(arg) for arg in types)
```

Comment:
This doesn't "smell" right to be to be creating a second temporary object and using Python iteration, but let's accept it for now.

I'll look into it more to see if what else we can squeeze out of this function in terms of performance, but right now let's just go for it.

Reply:
I could do:
```
        if type(types) is not tuple:)
            try:
                return self._lookup[(_to.parse(types),)]
            except KeyError:
                raise TypeError("specialisation not known for types: " + str(types)) from None
        types = tuple(_to.parse(arg) for arg in types)
        try:
            return self._lookup[types]
        except KeyError:
            raise TypeError("specialisation not known for types: " + str(types)) from None
```
But it also feels wrong...

Reply:
Yeah, I couldn't think of anything super neat off the top of my head either - I just wanted to bring it up.  I wouldn't worry about it right now because it's likely not much of an issue.  Just in case we revisit this in the future, the _potential_ problems are
 - `for x in y` Python iteration through a type that Cython may not be able to infer is `tuple` (should check the annotated Cython output / generated C++)
 - this requires allocation of a new `tuple`, in a tight, commonly called function, when we're sure at this point that we already have a tuple of the correct size created
 
Let's just leave it as-is for now - even looking back at my fairly recent Cython code I see parts where I could have done it better, so it's likely I'm just fretting over premature optimisation.